### PR TITLE
[runtime] context biasing with ac automaton

### PR DESCRIPTION
**Pull request by pengzhendong**

### CODE REVIEW

**Comment by pengzhendong** [comment]
librispeech:

old context graph: 8.14%(wer) 5.63%(u-wer) 30.19(b-wer)
ac automaton context graph: 8.13%(wer) 5.61%(u-wer) 30.26(b-wer)

aishell:

old context graph: 12.86%(含热词) 5.42%(不含热词)
ac automaton context graph: 12.99%(含热词) 5.42%(不含热词)

**Review by robin1001 — APPROVED** [review]

**Comment by yangyyt** [comment]
> librispeech: old context graph: 8.14%(wer) 5.63%(u-wer) 30.19(b-wer) ac automaton context graph: 8.13%(wer) 5.61%(u-wer) 30.26(b-wer)
> 
> aishell: old context graph: 12.86%(含热词) 5.42%(不含热词) ac automaton context graph: 12.99%(含热词) 5.42%(不含热词)

Can you provide some examples for us to try and learn from?

**Comment by kaixunhuang0** [comment]
> > librispeech: old context graph: 8.14%(wer) 5.63%(u-wer) 30.19(b-wer) ac automaton context graph: 8.13%(wer) 5.61%(u-wer) 30.26(b-wer)
> > aishell: old context graph: 12.86%(含热词) 5.42%(不含热词) ac automaton context graph: 12.99%(含热词) 5.42%(不含热词)
> 
> Can you provide some examples for us to try and learn from?

Context list of LibriSpeech can be obtained from here: https://github.com/facebookresearch/fbai-speech/tree/main/is21_deep_bias

**Comment by WjMessi1** [comment]
What is the maximum number of hot words now? When I test 125000 hot words, there is an error, can I change the maximum number of hot words?

File: runtime/core/kaldi/decoder/lattice-faster-decoder.cc [hunk 1]
```diff
@@ -933,28 +921,15 @@ void LatticeFasterDecoderTpl<FST, Token>::ProcessNonemitting(BaseFloat cutoff) {
...
-            if (arc.olabel == 0) {
-              e_new->val->context_state = tok->context_state;
-            } else {
-              e_new->val->context_state = context_graph_->GetNextState(
-                  tok->context_state, arc.olabel, &context_score,
-                  &is_start_boundary, &is_end_boundary);
-              graph_cost -= context_score;
-            }
+          if (context_graph_ != nullptr && changed) {
+            e_new->val->context_state = tok->context_state;
```
Comment (fclearner) [inline runtime/core/kaldi/decoder/lattice-faster-decoder.cc:928]: 东哥，请教下，processnonemitting这里为什么要把热词增强这部分代码去掉，有点没看懂

**Comment by fclearner** [comment]
@kaixunhuang0 大佬，请教个问题，v3.0.0的wenet fst热词对语言模型有要求吗，需要基于单字建模吗，我看热词处理的时候是会处理成单字的；我现在碰到个奇怪的问题，测了几条音频，新版的runtime(v3.0.0)热词权重给到20也没法生效；v2.2.1的runtime热词权重给到5就已经生效了，我这边啥都没改，怀疑我的参数传的有问题(两个版本用的同一套参数)，新版是不是有一些变动

**Comment by fclearner** [comment]
> @kaixunhuang0 大佬，请教个问题，v3.0.0的wenet fst热词对语言模型有要求吗，需要基于单字建模吗，我看热词处理的时候是会处理成单字的；我现在碰到个奇怪的问题，测了几条音频，新版的runtime(v3.0.0)热词权重给到20也没法生效；v2.2.1的runtime热词权重给到5就已经生效了，我这边啥都没改，怀疑我的参数传的有问题(两个版本用的同一套参数)，新版是不是有一些变动

大佬没事了，果然是语言模型构图的词表问题，我之前用的词，现在用字效果正常了，具体原因我再去看下代码，感谢

**Pull request merged**